### DISCUSSION TITLE
Callback for SOF

### DISCUSSION THREAD

**Discussion by 7Gluk** [hathach/tinyusb#1675] [Q&A]
Does the library have a callback for SOF? What should I do to use it?
If it's important, I'm making a composite HID device on STM32F103. And STM32 programming is new for me, I used to program AVR.

**Comment by hathach** [comment]
SOF only available for audio driver to compute feedback. Why would you want to use SOF ?

**Reply by 7Gluk** [reply]
SOF is used in firmware. The firmware is not written by me, I only transfer it from AVR to STM32.

**Reply by 7Gluk** [reply]
Can I add SOF callback for HID driver? And how to do it?

**Reply by hathach** [reply]
sof is not enabled for hid, you will need to modify tinyusb stack to do that yourself.

**Comment by 7Gluk** [comment]
I'm trying to add SOF callback myself:
add usbd_sof_enable(0, true); to the end of hidd_init in hid_device.c
add TU_ATTR_WEAK void tud_sof_cb(uint8_t rhport, uint32_t frame_count); in usbd.h
cange .sof = NULL to .sof = tud_sof_cb for DRIVER_NAME("HID") in usbd.c
but it does not working...

Also I'm trying to add if (tud_sof_cb) tud_sof_cb(event.rhport, event.sof.frame_count); to case DCD_EVENT_SOF: in usbd.c
but it does not working too...

**Reply by hathach** [reply]
enable debug log and keep debugging, you are in the right direction I guess.

**Reply by 7Gluk** [reply]
How can i do this?
I use TinyUSB in STM32CubeIDE project.

**Reply by hathach** [reply]
check getting started for how to enable log with gnu make. For any other IDEs, you are on your own, since I don't use any.

**Comment by 7Gluk** [comment]
How should I configure USART for log out?
I can't use SWO because i have only st-link v2 mini without SWO pinout.
Unfortunately my st-link v2 (not mini) is broken...

**Reply by hathach** [reply]
if you use one of the supported boards and gnu make, it is already configured correctly. Otherwise you need to read stm specs + code to write it yourself.

**Reply by 7Gluk** [reply]
I mean how the parameters should be set for USART? Or does it not matter?

**Reply by hathach** [reply]
please check existing code for one of supported board for detail. uart is retargeted with nanolib _write()
